pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,309 @@
+        self.step_and_cost()
+
+    def get_su_n_operators(self, restriction):
+        r"""Get the 2x2 SU(N) operators. The dimension of the group is N^2-1.
+
+        Args:
+            restriction (qml.Hamiltonian): Restrict the lie gradient to a subalgebra.
+
+        Returns:
+            List of (N^2)x(N^2) numpy complex arrays and corresponding paulis words
+        """
+
+        operators = []
+        names = []
+        # construct the corresponding pennylane observables
+        wire_map = dict(zip(range(self.nqubits), range(self.nqubits)))
+        if restriction is None:
+            for ps in qml.grouping.pauli_group(self.nqubits):
+                operators.append(ps)
+                names.append(qml.grouping.pauli_word_to_string(ps, wire_map=wire_map))
+        else:
+            for ps in set(restriction.ops):
+                operators.append(ps)
+                names.append(qml.grouping.pauli_word_to_string(ps, wire_map=wire_map))
+
+        return operators, names
+
+    def get_omegas(self):
+        r"""Measure the coefficients of the Lie gradient with respect to a Pauli word basis
+
+        We want to calculate the components of the Lie gradient with respect to a Pauli word basis
+        For a Hamiltonian of the form :math:`H = \sum_i O_i`, this can be achieved by calculating
+
+        .. math:
+
+            \omega_{i,j} = \text{Tr}([\rho, O_i] P_j)
+
+        where :math:`P_j` is a Pauli word in the set of Pauli monomials on :math:`N` qubits.
+
+        Via the parameter shift rule, the commutator can be calculated as
+
+        .. math:
+
+            [\rho, O_i] = \frac{1}{2}(V(\pi/2) \rho V^\dag(\pi/2) - V(-\pi/2) \rho V^\dag(-\pi/2))
+
+        where :math:`V` is the unitary generated by the Pauli word :math:`V(\theta) = \exp{-i\theta P_j }`.
+
+        Returns:
+            Array of omegas for each direction in the Lie algebra.
+
+        """
+
+        obs_groupings, _ = qml.grouping.group_observables(self.observables, self.coeffs)
+        # get all circuits we need to calculate the coefficients
+        circuits = algebra_commutator(
+            self.circuit.qtape,
+            obs_groupings,
+            self.lie_algebra_basis_names,
+            self.nqubits,
+        )[0]
+        # For each observable O_i in the Hamiltonian, we have to calculate all Lie coefficients
+        omegas = np.zeros((len(self.coeffs), len(self.lie_algebra_basis_names)))
+        idx = 0
+
+        for circuit_plus, circuit_min in circuits:
+            out_plus = qml.execute(circuit_plus, self.circuit.device, gradient_fn=None)
+            out_min = qml.execute(circuit_min, self.circuit.device, gradient_fn=None)
+            # depending on the length of the grouped observable, store the omegas in the array
+            omegas[idx : idx + len(out_plus[0]), :] = 0.5 * (
+                np.array(out_plus).T - np.array(out_min).T
+            )
+            idx += len(out_plus[0])
+        return np.dot(self.coeffs, omegas)
```

Comment:
very minor, but since you're returning a scalar here, could you potentially avoid building a large `omegas` array, and instead simply perform the reduction within th efour loop?